cmd/otel-collector/main.go

```diff
@@ -67,31 +67,22 @@ var cmd = &cobra.Command{
 }
 
 func mergeVertices(ch chan *bkclient.SolveStatus) []Vertex {
-	vertexByID := map[string]*bkclient.Vertex{}
+	seenVertices := map[string]*bkclient.Vertex{}
 	vertices := []*bkclient.Vertex{}
 	for msg := range ch {
 		if msg == nil {
 			break
 		}
 		for _, v := range msg.Vertexes {
-			vertex := vertexByID[v.Digest.String()]
-			if vertex == nil {
-				vertex = v
-				vertexByID[v.Digest.String()] = v
-				vertices = append(vertices, v)
+			if v.Completed == nil {
+				continue
```

Comment:
I'm not at all sure, but I'm wondering if this would mean that if you stopped a build suddenly then you'd never get any data for it.

If so, I could imagine that being annoying if you are trying dealing with a long hang of something, cancel the build, but then can't actually see what was taking so long.

Or maybe buildkit flushes `Completed`/`Canceled` progress events whenever you stop the build, in which case this is a non-issue. Worth a double check if you haven't already.

Reply:
OK, checking now how a cancelled build behaves.

Reply:
Here is the `Ctrl+C` interrupted run of `./hack/make engine:test`:

![image](https://user-images.githubusercontent.com/3342/216170786-72191cf2-fc1f-410f-aeeb-40574aaa4870.png)

This is what the trace looks like:

<img width="1726" alt="image" src="https://user-images.githubusercontent.com/3342/216170560-3790c037-4a20-41ec-9106-aa6cbfff3ef8.png">

And this is what the metrics look like:

<img width="1281" alt="image" src="https://user-images.githubusercontent.com/3342/216170380-a256f15c-acaf-4361-88f7-2c92de2e5f38.png">

Which makes me believe that we are good. Let's re-open and keep digging into this if you think otherwise.

Reply:
Awesome, that sounds good. I am gonna guess it's possible if buildkit or the local session process shuts down ungracefully for the traces to not complete but I think that's obscure enough to ignore for the moment.